MRG Freesurfer coordinate frame tutorial

larsoner opened the pull request:
The idea is to help people gain fluency with the Freesurfer coordinate frame (since we use it extensively), as well as some stuff about relationships between MRI voxels and affines/MRI coordinate frame. It's not done but at least gets some content out there.

Also makes much nicer some of our info repr stuff (dig, chs entries) by ensuring that on read the constants become "named constants", so that instead of seeing this on `master`:
```
>>> import mne
>>> from pprint import pprint
>>> info = mne.io.read_info(mne.datasets.sample.data_path() + '/MEG/sample/sample_audvis_raw.fif')
>>> pprint(info['chs'][0])
{'cal': 3.1600000394149674e-09,
 'ch_name': 'MEG 0113',
 'coil_type': 3012,
 'coord_frame': 1 (FIFFV_COORD_DEVICE),
 'kind': 1,
 'loc': array([-0.1066    ,  0.0464    , -0.0604    , -0.0127    ,  0.0057    ,
       -0.99990302, -0.186801  , -0.98240298, -0.0033    , -0.98232698,
        0.18674099,  0.013541  ]),
 'logno': 113,
 'range': 0.00030517578125,
 'scanno': 1,
 'unit': 201,
 'unit_mul': 0}
```
You get:
```
>>> pprint(info['chs'][0])
{'cal': 3.1600000394149674e-09,
 'ch_name': 'MEG 0113',
 'coil_type': 3012 (FIFFV_COIL_VV_PLANAR_T1),
 'coord_frame': 1 (FIFFV_COORD_DEVICE),
 'kind': 1 (FIFFV_MEG_CH),
 'loc': array([-0.1066    ,  0.0464    , -0.0604    , -0.0127    ,  0.0057    ,
       -0.99990302, -0.186801  , -0.98240298, -0.0033    , -0.98232698,
        0.18674099,  0.013541  ]),
 'logno': 113,
 'range': 0.00030517578125,
 'scanno': 1,
 'unit': 201 (FIFF_UNIT_T_M),
 'unit_mul': 0 (FIFF_UNITM_NONE)}
```

Todo:

- [x] Better explanation of Freesurfer surface RAS (how it's defined)
- [x] Surfaces (read in, show, units)
- [x] Surface alignment (fsave, fsave sph, sample sph, sample)
- [x] Surface decimation (show sphere, etc.)
- [x] Link to forward tutorial or something that talks about vertex removal, probably a `.. warn` to take care of #7007 (modify the `morph` tutorials and errors also?)
- [ ] ~~Source space entries (use_tris, etc.)~~

larsoner commented:
So far:

https://19221-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_freesurfer_coordinate_frames.html

drammock commented:
@larsoner I've pushed an expansion of the freesurfer tutorial, mostly just adding prose to the explanations without changing the code very substantively.  I feel like it ends kind of abruptly, not sure if there's a good way of summarizing / wrapping up / adding some "see also" text?

larsoner commented:
I updated the top comment with other stuff I plan to add, WDYT?

drammock commented:
TODO list seems reasonable.  I added one item: explanation of how "freesurfer surface RAS" is defined. So far we only say "it's different from nibabel's RAS, and the transform from voxels is called..." without ever saying how it's different. I tried reading the Freesurfer PPT slides that allegedly explain it, but if the answer is in there I was not smart/patient enough to extract it.

larsoner commented:
Yes, that has been my experience as well :)

I'll have to read and/or talk more to figure that out, but I agree it's worth doing...

hoechenberger commented on tutorials/source-modeling/plot_freesurfer_coordinate_frames.py:
```diff
@@ -0,0 +1,201 @@
+        text.set_path_effects([
+            path_effects.Stroke(linewidth=2, foreground='black'),
+            path_effects.Normal()])
+    # reorient for our data orientation and label
```
I think this needs more explanation. Why and how did you decide to reorient? Maybe one should add examples of

- `nib.io_orientation()`
- `nib.orientations.aff2axcodes()`
- `nib.as_closest_canonical()`

before moving on to this plotting example?

larsoner replied:
The logic here is specific to our known orientation, a variant of the general logic lives in https://github.com/mne-tools/mne-python/pull/7725. I agree this should be updated to use it

wmvanvliet commented on tutorials/source-modeling/plot_freesurfer_coordinate_frames.py:
```diff
@@ -0,0 +1,201 @@
+
+fiducials = mne.coreg.get_mni_fiducials('sample', subjects_dir=subjects_dir)
+nasion_mri = [d for d in fiducials if d['ident'] == FIFF.FIFFV_POINT_NASION][0]
+print(nasion_mri)  # note it's in MRI coords
```
what are "MRI coords"? Is that ijk? Is it Freesurfer RAS?

wmvanvliet commented on tutorials/source-modeling/plot_freesurfer_coordinate_frames.py:
```diff
@@ -0,0 +1,201 @@
+
+###############################################################################
+# We can also take the digitization point from the MEG data, which is in the
+# "head" coordinate frame:
```
Link to https://mne.tools/stable/overview/implementation.html#the-head-and-device-coordinate-systems?

wmvanvliet commented on tutorials/source-modeling/plot_freesurfer_coordinate_frames.py:
```diff
@@ -0,0 +1,201 @@
+print(nasion_head)  # note it's in "head" coordinates
+
+###############################################################################
+# To convert from head coordinate frame to MRI, we first apply the transform
```
to MRI surface RAS?

larsoner commented:
@wmvanvliet @hoechenberger I think I addressed your comments with the latest commit. (Had to force push to deal with a rebase problem)

hoechenberger commented:
Built documentation: https://21054-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_freesurfer_coordinate_frames.html#sphx-glr-auto-tutorials-source-modeling-plot-freesurfer-coordinate-frames-py

susamerz commented:
I like the hands on, step by step approach of the rest of the tutorial, but the following sentence had my (slightly sleep deprived) brain run out of processing capacity, even though I know the underlying theory:

" These data are voxel intensity values (unsigned integers in the range 0-255). A value data[i, j, k] at a given index triplet (i, j, k) is located at real-world (x, y, z) positions (in RAS orientation) in the scanner’s native coordinate frame (defined during acquisition) by the image’s affine transformation"

Trying to parse this I get:
These data are voxel intensity values (unsigned integers in the range 0-255). A value data[i, j, k] at a given index triplet (i, j, k) is located at real-world (x, y, z) positions (in RAS orientation).
The x y and z positions are in the scanners coordinate frame, which is defined during acquisition. 
The relationship between the [i,j,k] index and the x, y, z coordinates is saved (or written?) in the images affine transformation. 

Is this interpretation the intended one? If yes, feel free to use that wording, if no, please clarify :)

wmvanvliet commented:
I agree with @susamerz's comment. Otherwise, I'm very happy with this, great work @larsoner!

larsoner commented:
Okay addressed comments hopefully, @susamerz can you look again at the description of the affine?

I also added the rest of the missing content. @wmvanvliet @hoechenberger want to take another look to see if the new sections make sense to you?

https://21085-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_background_freesurfer_mne.html

@GuillaumeFavelier can you look at the 3D plotting bits and advise how to avoid using a private function? Maybe we need to expose how to get a renderer? Also take a look at the changes I needed to make to the renderer code in general, it was necessary to:

1. get things to plot the same in `mayavi` and `PyVista`
2. allow wireframe rendering (needed for this tutorial)

Once the 3D plotting issues are taken care of this will be ready for review/merge.

hoechenberger commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+#
+# Using FreeSurfer's surface reconstructions
+# ==========================================
+# An import part of what FreeSurfer does is provide cortical surface
```
```suggestion
# An important part of what FreeSurfer does is provide cortical surface
```

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+# We can also take the digitization point from the MEG data, which is in the
+# "head" coordinate frame.
+#
+# .. note::
+#      The head coordinate frame in MNE is the "Neuromag" head coordinate
+#      frame. The origin is given by the intersection between a line connecting
+#      the LPA and RPA and the line orthogonal to it that runs through the
+#      nasion. It is also in RAS orientation, meaning that +X runs through
+#      the RPA, +Y goes through the nasion, and +Z is orthogonal to these
+#      pointing upward. See :ref:`coordinate_systems` for more information.
```
This info is important enough that I don't see why it's called out as a note.  I'd integrate it with the sentence that precedes it.

larsoner replied:
I recently learned that `.. note` is supposed to be for calling out important information

https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#notes

I guess we could / should adopt this practice, and use `.. sidebar` for things that are asides / less important

drammock replied:
hmm.  The [official rST docs](https://docutils.sourceforge.io/docs/ref/rst/directives.html#note) don't actually give guidance on the semantics of the different admonition types, but it looks like [the sphinx docs do](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-note). TIL.  I guess we could argue about how applicable that guidance is for *tutorials* as opposed to *APIs*, but it doesn't seem important enough to bother; it's probably better to decide on guidelines that make sense for our project, document them, and stick to them.

larsoner replied:
I would vote for `.. note` as being for important things, `.. sidebar` for things that are less important / asides, and `.. warning` for things that will likely trip up users. I'm not sure the extent to which we have used `.. note` this way previously or not

drammock replied:
Some observations:

- IMO the main text of a docstring or tutorial ought to contain important information, so saying that `.. note::` is for "important things" doesn't actually clarify what should go in a note and what should be part of the main text.
- we've had complaints (I think from @jasmainak) about too many notes breaking up the flow of the page and making it harder to read.  I am generally sympathetic with that view.

WDYT about demote notes to be semantically like sidebars?  Maybe we could get by with only one of them, or maybe we would want to only use `.. note::` in docstrings (so they render inline), and only using `.. sidebar::` in examples/tutorials (where the generally longer prose passages makes sidebars look decent)?  We could define both of them as "an aside, containing useful information that is not strictly needed to use the function / follow the tutorial".

larsoner replied:
This one can be a sidebar because the focus is not really on the head coord frame. But the one about "MRI coordinates" in MNE really is important. I'll just make that one bold for now.

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+               d['ident'] == FIFF.FIFFV_POINT_NASION][0]
+print(nasion_head)  # note it's in "head" coordinates
+
+###############################################################################
```
```suggestion
###############################################################################
# Notice that in "head" coordinate frame the nasion has values of 0 for the
# ``x`` and ``z`` directions (which makes sense given that the nasion is used
# to define the ``y`` axis in that system).
```

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+print(nasion_head)  # note it's in "head" coordinates
+
+###############################################################################
+# To convert from head coordinate frame to MRI, we first apply the transform
```
```suggestion
# To convert from head coordinate frame to voxels, we first apply the head →
# MRI (surface RAS) transform
```

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+           'Nasion transformed from digitization')
+
+###############################################################################
+# Under the hood, in functions like :func:`mne.setup_source_space`,
```
```suggestion
# Under the hood, functions like :func:`mne.setup_source_space`,
```

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+# Using FreeSurfer's surface reconstructions
+# ==========================================
+# An important part of what FreeSurfer does is provide cortical surface
+# reconstructions. For example, let's read in and look at the ``white`` surface
```
```suggestion
# reconstructions. For example, let's load and view the ``white`` surface
```

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+# ==========================================
+# An important part of what FreeSurfer does is provide cortical surface
+# reconstructions. For example, let's read in and look at the ``white`` surface
+# of the brain. This is just a 3D mesh defined by a set of vertices (rr) with
+# shape ``(n_vertices, 3)`` and tris defining the triangles connecting the
+# vertices with shape ``(n_tris, 3)``.
```
```suggestion
# of the brain. This is a 3D mesh defined by a set of vertices (conventionally
# called ``rr``) with shape ``(n_vertices, 3)`` and a set of triangles
# (``tris``) with shape ``(n_tris, 3)`` defining which vertices in ``rr`` form
# each triangular facet of the mesh.
```

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+renderer.show()
+
+###############################################################################
+# We can also plot this on top of the MRI from before. These surfaces are
+# in millimeters in the MRI (FreeSurfer surface RAS) coordinate frame, so
+# let's get them to voxels by applying the inverse of the Torig (vox_mri_t):
```
```suggestion
# We can also plot the mesh on top of an MRI slice. The mesh surfaces are
# defined in millimeters in the MRI (FreeSurfer surface RAS) coordinate frame,
# so we can convert them to voxels by applying the inverse of the ``Torig``
# transform:
```

drammock commented on tutorials/source-modeling/plot_background_freesurfer_mne.py:
```diff
@@ -0,0 +1,425 @@
+
+###############################################################################
+# You can see that the fsaverage (purple) mesh is uniformly spaced, and the
+# sample (cyan) mesh has been deformed along the spherical surface by
```
```suggestion
# mesh for subject "sample" (in cyan) has been deformed along the spherical
# surface by
```

hoechenberger commented:
Awesome!

wmvanvliet commented:
cool!
